scipy/optimize/_chandrupatla.py

```diff
@@ -1,7 +1,16 @@
+import math
 import numpy as np
 from ._zeros_py import _rtol
 import scipy._lib._elementwise_iterative_method as eim
 from scipy._lib._util import _RichResult
+from scipy._lib._array_api import xp_clip, xp_minimum, xp_sign
+
+# TODO:
+# - debug test_maxiter_callback
+# - debug torch failure in test_special_cases - `float64` is fine?
+# - (maybe?) don't use fancy indexing assignment
+# - figure out how to replace the new `try`/`except`s
```

Comment:
It's not strictly correct, I don't think. The try/excepts are used to identify compressible arrays - the ones that should contain only "active" elements. But I can envision a case in which there is an attribute that is stored as a 1D array that *happens* to have the same size as the mask and gets compressed even though it's not supposed to. I need to investigate more closely. It would probably be better for the author using the framework to specify the attribute names of compressible arrays.

Reply:
I thought about it more carefully, and I think this is fine actually. I fixed the tests that I was having trouble with, so if tests pass, I think this is good from my side.